### ISSUE TITLE
Where is the context.pbtxt in UD_language?

### ISSUE THREAD

**Issue by aluminumbox** [dsindex/syntaxnet#34]
Hi @dsindex ,
I altered CORPUS_DIR=${CDIR}/UD_English to CORPUS_DIR=${CDIR}/corpus/ud-treebanks-v2.0/UD_Chinese. I thought this can train the Chinese corpus.
When I run $ ./train.sh -v -v, It says there is can not find context.pbtxt in UD_Chinese directory.
I am kind of confused. You have a directory named English, which contains the filed obtained from training. Also, you have context.pbtxt in UD_English. How can I run the train.sh without the context.pbtxt in UD_Chinese?
Am I misunderstanding something in here?

**Comment by dsindex** [comment]
hi~ 

there is an reference directory ‘UD_English’.

https://github.com/dsindex/syntaxnet/tree/master/UD_English

and you should modify the ‘train.sh’ script.
- copy needed files from UD_English to UD_chinese
- modify pathes for context.pbtxt

p.s. 
‘English’ directory is just copied version of trained model.

**Comment by aluminumbox** [comment]
Thank you @dsindex . 
I think you must be familiar with Parsey Universal. https://github.com/tensorflow/models/blob/master/research/syntaxnet/g3doc/universal.md.
I downloaded the Chinese model. It has:

char-map		last-word-map		parser-params		tagger-params
char-ngram-map		morph-label-set		prefix-table		tokenizer-params
checkpoint		morpher-params		suffix-table		tokenizer-params.meta
label-map		morphology-map		tag-map			word-map

However, it doesn't have context.pbtxt in it. Doesn't pretrained model means that it should be ready to use?  From your comments in other issues, It seems that pretrained model also need to be trained again and export.  I checked that your English directory. It is same as that of English model in parsey universal, except that you have context.pbtxt, which can only be obtained after training? In parsey universal, different models seems have different files. How did you trained it to get different context.pbtxt and exactly the same result as parsey universal?

I am really confused by this context.pbtxt,  why is pretrained model not ready to use and if so, how can I set the right parameter to achieve the accuaracy...
Looking forward for your reply~

**Comment by dsindex** [comment]
@aluminumbox 

the pre-trained model(Parsey Univeral) only have model files except context.pbtxt file.
but old version of Syntaxnet has context.pbtxt file. so, i copied it to 'English' directory.

```
commit c15fada28113eca32dc98d6e3bec4755d0d5b4c2
Author: Neal Wu <neal@nealwu.com>
Date:   Mon Mar 27 16:28:25 2017 -0700

$ ls
parsey_mcparseface  parsey_universal
$ pwd
~/syntaxnet/syntaxnet/models
$ cd parsey_universal/
$ ls
context-tokenize-zh.pbtxt  context.pbtxt  parse.sh  tokenize.sh  tokenize_zh.sh
```

as you concerned, i am not sure the parameters differ from languages or not.
however, 'English/context.pbtxt' was copied from the above example directory for Chinese. 
i guess it would work fine. 

if you want to use the pre-trained model of Chinese, then just copy 'English/context.pbtxt' to 'Chinese/context.pbtxt'. 

i made a test script for it named 'parse.sh'.

**Comment by aluminumbox** [comment]
Thank you @dsindex 
Oh I see. So, maybe the context.pbtxt in models/parsey_universal is for all pretrained models. I tried it ealier. It works, but makes some obvious mistakes.
Also, different language models have different files. For example, Russian model do not have char-map. I guess there maybe a context.pbtxt for each unique language model.

One more question. You said that English directory is just the downloaded model with context.pbtxt from old version, right? My Mac is too slow, could you tell me what I will obtain after I train the UD_English corpus?

**Comment by dsindex** [comment]
aha~ that would be big concern! ;;

**Comment by aluminumbox** [comment]
@dsindex In UD_English/context.pbtxt,  
Parameter {
  name: 'brain_parser_embedding_dims'
  value: '64;32;32'
}
In English/Context.pbtxt:
Parameter {
  name: "brain_parser_embedding_dims"
  value: "32;32;32;64"
}
So, I guess I will get a totally different context.pbtxt after the training using the setting in UD_English/context.pbtxt?

**Comment by dsindex** [comment]
yes, you are right.
the settings in UD_English and English(pretrained) are different.
for same setting, you should modify it;

**Comment by aluminumbox** [comment]
@dsindex 
Some thoughts. Most of the pre-trained language model contains same name files, while some like Russia miss some files, some like Chinese have extra files. I guess context.pbtxt is like the default parameters, and those files like char-map are generated automatically during training. As to why different language models contains different files, I'll try to train it. Maybe it is due to the language difference. Please correct me if I were wrong. Thank you for your help~

**Comment by aluminumbox** [comment]
Hi @dsindex 
I was reading your code to get my own parser trainer. You mentioned earlier that context.pbtxt in English directory was copied from syntaxnet repository. But in the training, the script used context.pbtxt in UD_English directory. Where did you get this context.pbtxt? I noticed that it is not the same as any context.pbtxt in syntaxnet model repository.

**Comment by dsindex** [comment]
i copied it(UD_English) from old version of repository.

checkout 

commit c15fada28113eca32dc98d6e3bec4755d0d5b4c2
Author: Neal Wu <neal@nealwu.com>
Date:   Mon Mar 27 16:28:25 2017 -0700

$ ls
parsey_mcparseface  parsey_universal
$ pwd
~/syntaxnet/syntaxnet/models
$ cd parsey_universal/
$ ls
context-tokenize-zh.pbtxt  context.pbtxt  parse.sh  tokenize.sh  tokenize_zh.sh

and i am not sure about
context.pbtxt in English directory. maybe old version of syntaxnet repo has it(it was not a model trained by myself)

**Comment by aluminumbox** [comment]
I checkout that commit. The context.pbtxt in parsey_universal is same as context.pbtxt in your English directory, not the one in UD_English directory, which is used in train.sh. Was that you misremembered this, or I made some stupid mistake?...

**Comment by dsindex** [comment]
ou~ i am sorry.
it seems that i didn’t remember exactly where the context.pbtxt comes from.

as you said, if the one in English dir is same as in the parsey_universal, that is copied version. 

and i am not sure for context.pbtxt in UD_English.

**Comment by aluminumbox** [comment]
Its okay. I used syntaxnet/context.pbtxt during the training and it works. I thought maybe you have some good hyperparameters in that particular context.pbtxt in UD_English directory. Thanks!
